modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C

```diff
@@ -0,0 +1,242 @@
+  std::vector<const Node *> neighbors;
+  MeshTools::find_nodal_neighbors(_mesh, node, _nodes_to_elem_map, neighbors);
+  for (const auto & n : neighbors)
+    Moose::out << " " << n->unique_id();
+  Moose::out << "\n";
+
+  // This is fine:
+  // return _u_nodal->getNodalValue(node);
+  // This is more convoluted, but it is used below in pPlus, etc
+  dof_id_type dof = node.dof_number(_u_nodal->sys().number(), _u_var_num, 0);
+  const NumericVector<Real> & u = *_u_nodal->sys().currentSolution();
+  return u(dof);
+}
+*/
+
+Real
+AdvectiveFluxCalculator::rPlus(const Node & node_i) const
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+  const Real p = PQPlusMinus(node_i, PQPlusMinusEnum::PPlus);
+  if (p == 0.0)
+    // Comment after Eqn (49): if P=0 then there's no antidiffusion, so no need to remove it
+    return 1.0;
+  const Real q = PQPlusMinus(node_i, PQPlusMinusEnum::QPlus);
+  const Real r = q / p;
+  return limitFlux(1.0, r);
+}
+
+Real
+AdvectiveFluxCalculator::rMinus(const Node & node_i) const
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+  const Real p = PQPlusMinus(node_i, PQPlusMinusEnum::PMinus);
+  if (p == 0.0)
+    // Comment after Eqn (49): if P=0 then there's no antidiffusion, so no need to remove it
+    return 1.0;
+  const Real q = PQPlusMinus(node_i, PQPlusMinusEnum::QMinus);
+  const Real r = q / p;
+  return limitFlux(1.0, r);
+}
+
+Real
+AdvectiveFluxCalculator::limitFlux(Real a, Real b) const
+{
+  if (_flux_limiter_type == FluxLimiterTypeEnum::None)
+    return 0.0;
+
+  if ((a >= 0.0 && b <= 0.0) || (a <= 0.0 && b >= 0.0))
+    return 0.0;
+  const Real s = (a > 0.0 ? 1.0 : -1.0);
+
+  const Real lal = std::abs(a);
+  const Real lbl = std::abs(b);
+  switch (_flux_limiter_type)
+  {
+    case FluxLimiterTypeEnum::MinMod:
+      return s * std::min(lal, lbl);
+    case FluxLimiterTypeEnum::VanLeer:
+      return s * 2 * lal * lbl / (lal + lbl);
+    case FluxLimiterTypeEnum::MC:
+      return s * std::min(0.5 * std::abs(a + b), 2.0 * std::min(lal, lbl));
+    case FluxLimiterTypeEnum::superbee:
+      return s * std::max(std::min(2.0 * lal, lbl), std::min(lal, 2.0 * lbl));
+    default:
+      return 0.0;
+  }
+}
+
+void
+AdvectiveFluxCalculator::pre()
+{
+  // Following algorithm sets _kij to zero for all nodes we can see
+
+  _kij = {};
+  // QUERY:: does the following loop run through nodes on this processor only (+ ghosted ones?) or
+  // all nodes?  We only want nodes on this processor
+  for (const auto & node_i : _mesh.getMesh().node_ptr_range())
+  {
+    std::map<Node, Real> nodal_flux = {};
+    std::vector<const Node *> neighbors;
+    MeshTools::find_nodal_neighbors(_mesh, *node_i, _nodes_to_elem_map, neighbors);
+    for (const auto & n : neighbors)
+      nodal_flux.emplace(*n, 0.0);
+    nodal_flux.emplace(*node_i, 0.0); // include node_i - node_i connection
+    _kij.emplace(*node_i, nodal_flux);
+  }
```

Comment:
Why would not you just traverse through `_kij` and set its elements to zero? Are you trying to preallocate the data structure? Because you could build up the data structure on the first pass and the subsequent ones would just zero it out. Just an idea.

Reply:
If you want to stick with this though, look at `MooseMesh::localNodesBegin` and `MooseMesh::localNodesEnd`.